juju/model.py

```diff
@@ -43,32 +45,88 @@ def get_relation(self, relation_name, relation_id=None):
                 # errors, ideally integrating the error catching with Juju's mechanisms.
                 raise TooManyRelatedApps(relation_name, num_related, 1)
 
+    def _is_local_unit(self, unit_name):
+        return unit_name == self._local_unit_name
+
     def get_unit(self, unit_name):
-        return self._cache.get(Unit, unit_name)
+        return self._cache.get(CacheKey(Unit, unit_name), unit_name, self._is_local_unit(unit_name), self._backend, self._cache)
+
 
+class CacheKey:
+    def __init__(self, entity_type, *key_parts):
+        self.entity_type = entity_type
+        self.key_parts = key_parts
+
+    def __hash__(self):
+        return hash((self.entity_type, *self.key_parts))
 
 class ModelCache(weakref.WeakValueDictionary):
-    def get(self, entity_type, *args):
-        key = (entity_type,) + args
+    def get(self, key, *init_args):
         entity = super().get(key)
         if entity is None:
-            entity = entity_type(*args, cache=self)
+            entity = key.entity_type(*init_args)
             self[key] = entity
         return entity
 
+class EntityStatus:
+    """A helper class for getting and setting entity (application or unit) status."""
+
+    def __init__(self):
+        self._cached = None
+
+    def __get__(self, instance, owner):
+        is_app = isinstance(instance, Application)
```

Comment:
It doesn't make much sense to have a mixin class that actually knows about the exact details of all classes where it's used. It all becomes a single big and complex class that pretends to be a set of composible types.

Reply:
I wonder what the actual trade-off here is in duplication vs complexity with having `EntityStatus` as a separate "generic" descriptor class, vs just using `@property` in the `Application` and `Unit` classes?

Reply:
See #52